lib/declare_schema/model/habtm_model_shim.rb

```diff
@@ -5,28 +5,21 @@ module Model
     class HabtmModelShim
       class << self
         def from_reflection(refl)
-          join_table = refl.join_table
-          foreign_keys_and_classes = [
-            [refl.foreign_key.to_s, refl.active_record],
-            [refl.association_foreign_key.to_s, refl.class_name.constantize]
-          ].sort { |a, b| a.first <=> b.first }
-          foreign_keys = foreign_keys_and_classes.map(&:first)
-          foreign_key_classes = foreign_keys_and_classes.map(&:last)
-          # this may fail in weird ways if HABTM is running across two DB connections (assuming that's even supported)
-          # figure that anybody who sets THAT up can deal with their own migrations...
-          connection = refl.active_record.connection
-
-          new(join_table, foreign_keys, foreign_key_classes, connection)
+          new(refl.join_table, [refl.foreign_key, refl.association_foreign_key],
+                               [refl.active_record.table_name, refl.class_name.constantize.table_name])
         end
       end
 
-      attr_reader :join_table, :foreign_keys, :foreign_key_classes, :connection
+      attr_reader :join_table, :foreign_keys, :parent_table_names
 
-      def initialize(join_table, foreign_keys, foreign_key_classes, connection)
+      def initialize(join_table, foreign_keys, parent_table_names)
+        foreign_keys.is_a?(Array) && foreign_keys.size == 2 or
+          raise ArgumentError, "foreign_keys must be <Array[2]>; got #{foreign_keys.inspect}"
+        parent_table_names.is_a?(Array) && parent_table_names.size == 2 or
+          raise ArgumentError, "parent_table_names must be <Array[2]>; got #{parent_table_names.inspect}"
         @join_table = join_table
-        @foreign_keys = foreign_keys
-        @foreign_key_classes = foreign_key_classes
-        @connection = connection
```

Comment:
> Models are still expected to implement connection in a few places.

True. I'm just trying to chase away dependencies in general, when I can. Fewer implicit dependencies is always an improvement...unless it causes the explicit dependencies to get annoying.

It is possible to use `ActiveRecord::Base.connection`, but in the general case that may not work, because Rails allows each model to be stored in a different database and hence need a different connection. I don't think that's very commonly used, but it definitely does get some use. I'm also not certain how well `declare_schema` would work straddling multiple databases like.

Still, to answer this question:

>  Definitely would be interested in learning about the practical different between ActiveRecord::Base.connection and refl.active_record.connection.

The one in ActiveRecord::Base is the "main" connection. Models typically share that same connection pool, but they are allowed to specify a different connection spec (and hence connection pool) if they want to, so that a subset of the models might live in a separate DB. This support was added to Rails 6 and roughly encompasses what we've been doing with ActiveTableSet. So we think we can probably phase out ATS.

So best practice is:
1. Use the connection from the model you're accessing.
2. If you have general code that isn't tied to any particular model, use ActiveRecord::Base.

Reply:
Awesome thanks for that explanation that's really helpful.

So in this case, since the `HabtmModelShim` doesn't include `ActiveRecord::ConnectionHandling` we still need to copy/store that connection from the reflection right?

Reply:
@dcaddell Yes. In the latest feature branch I decided to [make that explicit, with a keyword argument](https://github.com/Invoca/declare_schema/pull/162/commits/257883c84d2180acea5f2be502c6ccb05b1965fb). I like how clear and non-error-prone that is. It was also really easy to TDD.